### PR TITLE
Add a basic frontend to the web plugin

### PR DESCRIPTION
**Pull request by tobim**
This PR adds a built-in web front end to the `web` plugin.

It can be opened at the root of the HTTP server.
For example with this in your `vast.yaml`:
```yaml
vast:
  plugins: [ "all" ]
  start:
    print-endpoint: true
    commands:
      - web server

plugins:
  web:
    mode: dev
    bind: 127.0.0.1
    port: 5657
```
You can point your browser to http://127.0.0.1:5657 to find the app.

### :memo: Reviewer Checklist

Review this pull request by ensuring the following items:

- [ ] All user-facing changes have changelog entries
- [ ] User-facing changes are reflected on [vast.io](https://github.com/tenzir/vast/tree/master/web)

### CODE REVIEW

**Comment by dit7ya** [comment]
Seems like `pnpm` needs to be installed in the CI environment? https://github.com/tenzir/vast/actions/runs/3265777637/jobs/5368525733#step:7:54

File: plugins/web/ui/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/web/ui/CMakeLists.txt:15]: You probably don't wanna build this in the source directory, but rather copy the sources over to the build directory beforehand.

Comment (tobim) [reply plugins/web/ui/CMakeLists.txt:15]: Why?

Comment (dominiklohmann) [reply plugins/web/ui/CMakeLists.txt:15]: Because `rm -rf path/to/build` should clean up everything that `cmake -B path/to/build && cmake --build path/to/build` produced. Currently this is producing build artifacts in the source directory.

**Comment by dit7ya** [comment]
`cmake -B build -DCMAKE_BUILD_TYPE=Release -DVAST_PLUGINS=plugins/web` fails with

```console
CMake Error at plugins/web/aux/CMakeLists.txt:20 (add_custom_command):
  add_custom_command Wrong syntax.  A TARGET or OUTPUT must be specified.
```

**Comment by lava** [comment]
@dit7ya This is caused by a missing bundled restinio (the error message is pretty terrible tbh), it should be solved by `git submodule update --init`.

**Comment by mavam** [comment]
There are two `vast.svg` files added in this PR. Is this on purpose?

File: plugins/web/ui/src/routes/status/+page.svelte
Comment (lava) [inline plugins/web/ui/src/routes/status/+page.svelte:17]: Is there a way to write this so that the "installed" version automatically uses its own domain as `API_BASE`, while still allowing development against an external server?

Comment (dit7ya) [reply plugins/web/ui/src/routes/status/+page.svelte:17]: Yes, that can be done. Such parameters are usually done via `.env` files. Do you want me to push a commit here?

Comment (lava) [reply plugins/web/ui/src/routes/status/+page.svelte:17]: Yes, that would be great

Comment (dit7ya) [reply plugins/web/ui/src/routes/status/+page.svelte]: Done. To configure the API Endpoint - copy the `.env.example` file to `.env` and edit the value.

Comment (netantho) [reply plugins/web/ui/src/routes/status/+page.svelte]: After trying for 30min, I didn't manage to make the `.env` work:
`pnpm exec vite dev` with `.env` in the same folder. I resorted to manually editing the svelte source file to make this work.

File: plugins/web/ui/src/routes/status/+page.svelte
Comment (netantho) [inline plugins/web/ui/src/routes/status/+page.svelte:72]: Typo: ch*e*cked

**Comment by netantho** [comment]
When I don't have any record in VAST, I get the following error:
```
bodyRows.js?v=76b13408:124 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'map')
    at getBodyRows (bodyRows.js?v=76b13408:124:23)
    at createViewModel.js?v=76b13408:12:16
    at sync (index.mjs:77:28)
    at index.mjs:95:9
    at Object.subscribe2 [as subscribe] (index.mjs:48:20)
    at subscribe (index.mjs:55:25)
    at index.mjs:85:62
    at Array.map (<anonymous>)
    at index.mjs:85:44
    at Object.subscribe2 [as subscribe] (index.mjs:48:20)
```

The UI shows the VAST server as not running where it's actually running, just in the CLI `vast count` returns `0` (try moving you're `vast.db` folder for example to get a new empty one)

I gets solved after importing a few records.

File: plugins/web/ui/src/routes/status/+page.svelte
Comment (dit7ya) [reply plugins/web/ui/src/routes/status/+page.svelte]: I am suspecting this is due to CORS issues. Did you see anything indicating that in the console? Anyway, I have pushed a commit that changes the endpoint from `127.0.0.1` to `localhost` which temporarily seems to fix the CORS issue. Paging @lava about the backend.

**Comment by dit7ya** [comment]
@netantho Thanks for the feedback. I have pushed commits which should fix them. Since the frontend is at the very basic state right now, a lot of things are yet to be implemented - which should prevent this kind of bugs.

**Comment by netantho** [comment]
> @netantho Thanks for the feedback. I have pushed commits which should fix them. Since the frontend is at the very basic state right now, a lot of things are yet to be implemented - which should prevent this kind of bugs.

Thanks @dit7ya! I confirm that your fixes for both of the bugs I found and reported :-)

File: plugins/web/ui/src/routes/status/+page.svelte
Comment (lava) [reply plugins/web/ui/src/routes/status/+page.svelte]: Hard to say without seeing the exact error, from my understanding of CORS the `Access-Control-Allow-Origin: *` header should allow embedding by arbitrary origins, but maybe this was against an older version of VAST or some browsers are more strict about cross-site requests than others?

**Comment by dit7ya** [comment]
I have figured out some issues in the CI for this PR and fixing them on https://github.com/tenzir/vast/pull/2656. I am going to push some of those changes, and also going to probably rebase this on latest master, if there is no objection.

File: plugins/web/ui/CMakeLists.txt
Comment (dominiklohmann) [reply plugins/web/ui/CMakeLists.txt]: This is resolved now.

File: Dockerfile
Comment (dominiklohmann) [inline Dockerfile:52]: This has to come before the `rm -rf /var/lib/apt/lists/*` in order to avoid inflating the generated image size.

File: cmake/VASTConfig.cmake.in
Comment (dominiklohmann) [inline cmake/VASTConfig.cmake.in:20]: This is unused and can be removed.

File: .github/workflows/style-check.yaml
Comment (dominiklohmann) [inline .github/workflows/style-check.yaml:24]: This must also include .cpp.in and .hpp.in files. Generally I think the blacklist approach is better than the whitelist approach, so I'd much prefer if we could exclude the web UI from this instead.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:549]: In all instances of `corepack enable` in the CI we should give the step a titlecased name for consistency with the other steps.

File: plugins/web/include/web/mime.hpp
Comment (dominiklohmann) [inline plugins/web/include/web/mime.hpp:17]: This needs to manually be added to VAST.spdx.

File: plugins/web/ui/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/web/ui/CMakeLists.txt:52]: `s/Builind UI to/Building UI in`, not sure what I was thinking here

File: plugins/web/ui/src/app.html
Comment (dominiklohmann) [inline plugins/web/ui/src/app.html:5]: I cannot comment on the SVG file directly, but there are two identical vast.svg files in this PR.

Additionally there's a favicon.png file that could probably also be the same vast.svg instead.

File: scripts/git-setup.sh
Comment (dominiklohmann) [inline scripts/git-setup.sh:59]: Same as above in the CI, this is incomplete.

File: scripts/git-setup.sh
Comment (dominiklohmann) [inline scripts/git-setup.sh:66]: Is this to support Git worktrees?

File: plugins/web/ui/src/routes/status/+page.svelte
Comment (dominiklohmann) [inline plugins/web/ui/src/routes/status/+page.svelte:2]: All of the frontend files use tabs rather than 2 spaces for indentation, which goes against our style guidelines.

**Review by dominiklohmann — COMMENTED** [review]
Except for the Nix fixes I think this is almost ready to go. I have some minor feedback that can be addressed while we work on fixing this with Nix in parallel.

File: plugins/web/ui/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/web/ui/CMakeLists.txt:3]: If anything we should need it only in the `file(COPY ... DESTINATION ...)`, but not in this variable, which we also use in other places that definitely do not like the trailing slash.

Comment (dit7ya) [reply plugins/web/ui/CMakeLists.txt:3]: Seems like this is what was creating the nested www directories. 🤷

File: scripts/git-setup.sh
Comment (dit7ya) [reply scripts/git-setup.sh:59]: Same here @tobim

File: scripts/git-setup.sh
Comment (dit7ya) [reply scripts/git-setup.sh:66]: @tobim

File: Dockerfile
Comment (tobim) [inline Dockerfile:48]: We can drop the pinning now.

File: flake.nix
Comment (tobim) [inline flake.nix:17]:
```suggestion
  outputs = { self, nixpkgs, flake-utils, ... }@inputs: {
```

File: nix/overlay.nix
Comment (tobim) [inline nix/overlay.nix:197]: RE: Not sure what the best approach is here, I think we basically want to exclude any build artifacts here, if possible without repeating the contents of `.gitignore`.

File: nix/vast/default.nix
Comment (tobim) [inline nix/vast/default.nix:77]: This should really no be necessary.

File: scripts/git-setup.sh
Comment (tobim) [reply scripts/git-setup.sh:66]: Yes!

File: nix/static-binary.sh
Comment (tobim) [inline nix/static-binary.sh:92]:
```suggestion
```

File: nix/vast/default.nix
Comment (tobim) [inline nix/vast/default.nix:37]:
```suggestion
, pkgsBuildHost
```

File: nix/vast/default.nix
Comment (tobim) [inline nix/vast/default.nix:120]:
```suggestion
    "-DVAST_WEB_UI_BUNDLE=${pkgsBuildHost.vast-ui}"
```

**Review by tobim — COMMENTED** [review]
I tested this locally and the static binary package now comes with a nice web ui! :partying_face: 

One extra note because you can't comment on files in GitHub:
This adds lock files for both yarn and pnpm, do we really need both?

File: scripts/git-setup.sh
Comment (tobim) [reply scripts/git-setup.sh:66]: Yes!

File: plugins/web/ui/CMakeLists.txt
Comment (dominiklohmann) [reply plugins/web/ui/CMakeLists.txt]: I implemented a more appropriate fix to this by appending the trailing slashes where required.

**Review by dominiklohmann — APPROVED** [review]
This is ready to ship in my opinion. I'd appreciate a second review from either @lava or @tobim, though, before merging this.

**Review by lava — APPROVED** [review]
From my side this is good to go as well but I went ahead and also merged #2681 into this PR so we atomically have the SBOM changes when we add the frontend code.

**Review by tobim — COMMENTED** [review]
This really deserves a changelog entry before it gets merged. @dit7ya do you want me to write one?

**Comment by dit7ya** [comment]
Yes, that would be great @tobim. 

On a sepatate note, macOS CI still seems unhappy 😭.

**Comment by tobim** [comment]
> Yes, that would be great @tobim.
> 
> On a sepatate note, macOS CI still seems unhappy sob.

I think this is an unrelated bug that somehow reached `master`: https://github.com/tenzir/vast/pull/2686#issuecomment-1310493648

**Comment by dominiklohmann** [comment]
I pushed a fix for that @tobim, see comment in that PR.

**Pull request merged**
